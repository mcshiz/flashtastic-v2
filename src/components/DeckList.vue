<template>
	<div v-if="!loadingDecks">
		<div v-if="Object.keys(decks).length === 0">
			<div class="col-12">
				<p>Create a deck</p>
				<create-new-deck-button></create-new-deck-button>
			</div>
		</div>
		<div class="col">
			<ul class="list-group">
				<li v-for='(deck, key) in decks' :key='`deck-${key}`' class="list-group-item">
					<div class="row">
						<div class="col-6 col-sm-8 col-md-10 text-left justify-content-center align-self-center">
							<b>{{deck.name}}</b>
							<br>
							{{Object.keys(deck.cards).length}} Cards in Deck
							<br>
							<span v-if="deck.scores"
									v-bind:class="{
													'text-success': highScore(deck) / Object.keys(deck.cards).length === 1,
													'text-warning': highScore(deck) / Object.keys(deck.cards).length >= 0.5 && highScore(deck) / Object.keys(deck.cards).length < 1,
													'text-danger': highScore(deck) / Object.keys(deck.cards).length <= 0.5}">
								High Score: <b>{{highScore(deck)}} / {{Object.keys(deck.cards).length}}</b>
							</span>
							<span v-else>Not Completed</span>
						</div>
						<div class="col-4 col-sm-2 col-md-1 pull-right align-self-center">
							<button class="btn btn-info btn-sm mb-1 col-12" v-on:click="loadDeck(deck, key)">Study</button>
						</div>
						<div v-if="(deck.permissions === 'private' || deck.creator === user.id)" class="col-2 col-sm-2 col-md-1 align-self-center">
							<button class="btn btn-default btn-sm col-12" v-on:click="editDeck(key)">Edit</button>
							<span v-on:click="deleteDeck(key)" class="delete-deck"><i class="fa fa-trash"></i></span>
						</div>
					</div>
				</li>
			</ul>
		</div>
	</div>
	<div v-else>Loading...</div>
</template>
<script>
import CreateNewDeckButton from './CreateNewDeckButton'
import { mapState } from 'vuex'
export default {
	components: {
		'create-new-deck-button': CreateNewDeckButton
	},
	name: 'DeckList',
	props: ['decks'],

	methods: {
		editDeck: function(key, permissions) {
			this.$router.push({name: 'EditDeck', params: {key: key, permissions: permissions}})
		},
		loadDeck: function(deck, key) {
			this.$router.push({name: 'Quiz', params: {key: key, name: deck.name, permissions: deck.permissions}})
		},
		deleteDeck: function(key) {
			if(window.confirm('Are you sure you want to delete this deck?')) {
				this.$store.dispatch('DELETE_DECK_BY_KEY', key)
			}
		},
		// fix this shit
		highScore: function(deck) {
			let highScore = 0
			for(let key in deck.scores) {
				if(deck.scores[key].correct > highScore) {
					highScore = deck.scores[key].correct
				}
			}
			return highScore
		}
	},
	computed: {
		...mapState(['loadingDecks', 'user'])
	}
}
</script>

<style>
	.delete-deck {
		line-height: 48px;
	}
</style>
