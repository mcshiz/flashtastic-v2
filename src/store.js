import Vue from 'vue'
import Vuex from 'vuex'
import fire from './fire'
import * as Firebase from 'firebase'
Vue.use(Vuex)

class BlankDeck {
	constructor(state) {
		this.name = ''
		this.cards = {}
		this.key = ''
		this.permissions = 'private'
		this.creator = state.user.id
	}
}
export default new Vuex.Store({
	state: {
		workingDeck: {},
		authenticated: false,
		user: {},
		privateDecks: {},
		publicDecks: {},
		loadingDecks: true,
		errorMessage: '',
		successMessage: '',
		successTitle: ''
	},
	actions: {
		INITIALIZE: ({commit, dispatch}) => {
			return new Promise((resolve, reject) => {
				let user
				let authd
				Firebase.auth().onAuthStateChanged((result) => {
					if (result !== null && result.uid) {
						authd = true
						user = {
							displayName: result.displayName,
							email: result.email,
							photoURL: result.photoURL,
							deckRef: `users/${result.uid}/decks`,
							scoreRef: `users/${result.uid}/scores`,
							storageRef: `users/${result.uid}/images`,
							id: result.uid
						}
					} else {
						user = {}
						authd = false
					}
					commit('SET_AUTHENTICATED', {authd, user})
					dispatch('LOAD_DECKS')
					resolve()
				})
			})
		},
		SIGN_IN: ({ commit, dispatch }) => {
			let provider = new Firebase.auth.GoogleAuthProvider()
			Firebase.auth().signInWithPopup(provider).then(() => {
				dispatch('INITIALIZE')
			}).catch(function(error) {
				commit('SET_ERROR_MESSAGE', error)
			})
		},
		SIGN_OUT: ({ commit }) => {
			Firebase.auth().signOut().then(function() {
				commit('SIGN_OUT_USER')
			}, function(error) {
				commit('SET_ERROR_MESSAGE', error)
			})
		},
		CREATE_BLANK_DECK: ({commit, state}) => {
			return new Promise((resolve, reject) => {
				let newDeck = new BlankDeck(state)
				commit('SET_WORKING_DECK', newDeck)
				resolve()
			})
		},
		LOAD_DECKS: ({ commit, state }) => {
			let publicDecks = {}
			let privateDecks = {}
			let decksRef = fire.database().ref('decks/').orderByKey().limitToLast(100)
			decksRef.on('value', snapshot => {
				let decks = snapshot.val()
				for(let key in decks) {
					if(decks[key].permissions === 'public') {
						publicDecks[key] = decks[key]
					} else if(decks[key].permissions === 'private' && decks[key].creator === state.user.id) {
						privateDecks[key] = decks[key]
					}
				}
				if(state.authenticated) {
					let scoresRef = fire.database().ref(`${state.user.scoreRef}`)
					scoresRef.on('value', snapshot => {
						for(let key in snapshot.val()) {
							if(decks[key]) {
								decks[key].score = snapshot.val()[key]
							}
						}
						commit('SET_LOADED_DECKS', {'private': privateDecks, 'public': publicDecks})
					})
				} else {
					commit('SET_LOADED_DECKS', {'private': privateDecks, 'public': publicDecks})
				}
			})
		},
		LOAD_DECK_BY_KEY: ({ commit, state }, {key, permissions}) => {
			return new Promise((resolve, reject) => {
				let decksRef = fire.database().ref(`decks/${key}`)
				decksRef.on('value', snapshot => {
					let deck = {
						...snapshot.val()
					}
					if(state.authenticated) {
						let scoresRef = fire.database().ref(`${state.user.scoreRef}/${key}`)
						scoresRef.on('value', snapshot => {
							deck.scores = snapshot.val()
							deck.key = snapshot.key
							commit('SET_WORKING_DECK', deck)
							resolve()
						})
					} else {
						commit('SET_WORKING_DECK', deck)
						resolve()
					}
				})
			})
		},
		UPDATE_DECK: ({commit, state}) => {
			return new Promise((resolve, reject) => {
				let decksRef = fire.database().ref(`decks/${state.workingDeck.key}`)
				decksRef.set(state.workingDeck).then(response => {
					resolve(response)
				})
			})
		},
		UPDATE_WORKING_DECK_IN_STATE: ({ commit }, deck) => {
			commit('SET_WORKING_DECK', deck)
		},
		DELETE_DECK_BY_KEY: ({ commit, state }, key) => {
			return new Promise((resolve, reject) => {
				// need to figure out a way to recursively delete files @ directory
				// Firebase.storage().ref().child(`${state.user.storageRef}/${key}`).delete() doesnt work
				// because we need to specify a ref all the way down to the exact filename
				fire.database().ref(`/decks/${key}`).remove().then(() => {
					fire.database().ref(`/decks/${key}`).remove().then(() => {
						resolve()
					})
				})
			})
		},
		SAVE_SCORE: ({ commit, state, dispatch }, score) => {
			let workingDeck = state.workingDeck.key
			let timestamp = new Date().getTime() * 1000
			fire.database().ref(`${state.user.scoreRef}/${workingDeck}/${timestamp}`).set({
				correct: score.correct,
				incorrect: score.incorrect
			})
				.catch(error => {
					commit('SET_ERROR_MESSAGE', error)
				})
		},
		UPLOAD_IMAGE: ({ commit, state }, {image, name}) => {
			let storageRef = Firebase.storage().ref()
			let filename = `${new Date().getTime() * 1000}_${name}`
			let newImageRef = storageRef.child(`/images/${filename}`)
			return new Promise((resolve, reject) => {
				newImageRef.put(image).then(snapshot => {
					if(snapshot.state === 'success') {
						resolve(snapshot)
					} else {
						reject(snapshot)
					}
				})
			})
		},
		CREATE_DECK: ({ commit, state }) => {
			return new Promise((resolve, reject) => {
				fire.database().ref('/decks').push(state.workingDeck)
					.then(() => {
						commit('SET_RESET')
						resolve()
					}).catch((error) => {
						commit('SET_ERROR_MESSAGE', error)
						reject(new Error(error))
					})
			})
		},
		SHOW_ERROR: ({ commit }, errorMessage) => {
			commit('SET_ERROR_MESSAGE', errorMessage)
		},
		RESET_ERROR: ({ commit }) => {
			commit('SET_ERROR_MESSAGE', '')
		},
		SHOW_SUCCESS: ({ commit }, {message, title}) => {
			commit('SET_SUCCESS_MESSAGE', {message, title})
		},
		RESET_SUCCESS: ({ commit }) => {
			commit('SET_SUCCESS_MESSAGE', {message: '', title: ''})
		}
	},
	mutations: {
		SET_SUCCESS_MESSAGE: (state, {message, title}) => {
			state.successMessage = message
			state.successTitle = title
		},
		SET_LOADED_DECKS: (state, decks) => {
			state.privateDecks = Object.assign({}, decks.private)
			state.publicDecks = Object.assign({}, decks.public)
			state.loadingDecks = false
		},
		SET_WORKING_DECK: (state, deck) => {
			state.workingDeck = Object.assign({}, state.workingDeck, deck)
		},
		SIGN_OUT_USER: (state) => {
			state.authenticated = false
			state.user = Object.assign({})
			state.myDecks = Object.assign({})
		},
		SET_AUTHENTICATED: (state, {authd, user}) => {
			state.authenticated = authd
			state.user = Object.assign({}, user)
		},
		SET_EDITS: (state, { index, card }) => {
			function replaceAt(array, index, value) {
				const ret = array.slice(0)
				ret[index] = value
				return ret
			}
			state.newDeck.questions = replaceAt(state.newDeck.questions, index, card)
		},
		SET_ERROR_MESSAGE: (state, errorMessage) => {
			state.errorMessage = errorMessage
		},
		SET_RESET: (state) => {
			state.newDeck = Object.assign({}, new BlankDeck(state))
		}
	}
})
