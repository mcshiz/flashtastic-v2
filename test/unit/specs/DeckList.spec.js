import { mount, shallow } from '@vue/test-utils'
import Vue from 'vue/dist/vue.common.js'
import Vuex from 'vuex'
import store from '../../../src/store'
import DeckList from '@/components/DeckList'

Vue.use(Vuex)

describe('Items.vue', () => {
	it('Loads Initial Data', (done) => {
		store._actions.LOAD_MY_DECKS = function() {
			commit('SET_MY_DECKS', {decks: {
					"0" : {
						"id" : 0,
						"image" : "/static/planets.jpeg",
						"name" : "Planets of the apes",
						"questions" : [ {
							"answer" : "Mercury",
							"id" : 0,
							"question" : "Planet closest to the sun",
							"result" : ""
						}, {
							"answer" : "Neptune",
							"id" : 1,
							"question" : "Planet furthest away from the sun",
							"result" : ""
						}, {
							"answer" : "Jupiter",
							"id" : 2,
							"question" : "What is the largest planet",
							"result" : ""
						}, {
							"answer" : "Venus",
							"id" : 3,
							"question" : "What is the hottest planet",
							"result" : ""
						}, {
							"answer" : "Mercury",
							"id" : 4,
							"question" : "What is the smallest planet",
							"result" : ""
						} ]
					},
					"1" : {
						"id" : 1,
						"image" : "/static/chemistry.jpg",
						"name" : "Chemistry",
						"questions" : [ {
							"answer" : "Hydrogen",
							"id" : 0,
							"question" : "H",
							"result" : ""
						}, {
							"answer" : "Lead",
							"id" : 1,
							"question" : "Pb",
							"result" : ""
						}, {
							"answer" : "Oxygen",
							"id" : 2,
							"question" : "Element number 8",
							"result" : ""
						}, {
							"answer" : "Iron",
							"id" : 3,
							"question" : "Fe",
							"result" : ""
						}, {
							"answer" : "Not Lead",
							"id" : 4,
							"question" : "What is the lightest element",
							"result" : ""
						} ]
					},
					"2" : {
						"id" : 2,
						"image" : "/static/trees.jpg",
						"name" : "Trees",
						"questions" : [ {
							"answer" : "Giant Sequoia",
							"id" : 0,
							"question" : "What is the name of the largest tree species",
							"result" : ""
						}, {
							"answer" : "Bristlecone Pine   ",
							"id" : 1,
							"question" : "What is the name of the oldest tree species",
							"result" : ""
						}, {
							"answer" : "Dendrochronology",
							"id" : 2,
							"question" : "What is the name for dating and studying trees",
							"result" : ""
						}, {
							"answer" : "False",
							"id" : 3,
							"question" : "True of False: Trees die of old age",
							"result" : ""
						}, {
							"answer" : "In their Cones",
							"id" : 4,
							"question" : "How do pine tress spread seeds",
							"result" : ""
						} ]
					},
					"-L3Ms2pFdC0YqgQPoJ2E" : {
						"id" : 0,
						"name" : "firebase",
						"questions" : [ {
							"answer" : "blue",
							"id" : 0,
							"question" : "what color is the sky",
							"result" : ""
						}, {
							"answer" : "Round",
							"id" : 1,
							"question" : "What shape is the earth",
							"result" : ""
						} ]
					}
				}})
		}
		const vm = new Vue({
			template: '<div><test></test></div>',
			store: store,
			components: {
				'test': DeckList
			}
		}).$mount()
		Vue.nextTick()
			.then(() => {
				expect(Object.keys(store.state.decks).length).toBe(4)
				done()
			})
			.catch(done)
	})
	it('Renders a list', (done) => {
		const vm = new Vue({
			template: '<div><test></test></div>',
			store: store,
			components: {
				'test': DeckList
			}
		}).$mount()
		Vue.nextTick()
			.then(() => {
				expect(vm.$el.querySelectorAll('.list-group-item').length).toEqual(4)
				done()
			})
			.catch(done)
	})
	it('Matches its snapshot', (done) => {
		const vm = new Vue({
			template: '<div><test></test></div>',
			store: store,
			components: {
				'test': DeckList
			}
		}).$mount()
		Vue.nextTick()
			.then(() => {
				expect(vm.$el).toMatchSnapshot()
				done()
			})
			.catch(done)
	})
})
