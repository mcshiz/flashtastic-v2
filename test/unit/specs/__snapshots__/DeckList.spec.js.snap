// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items.vue Matches its snapshot 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
